### PR TITLE
Delete exceeding versions when adding a new alias

### PR DESCRIPTION
**Pull request by AymericDu**
See #1163

### CODE REVIEW

File: tests/functional/content/test_content_versioning.py [hunk 1]
```diff
@@ -63,3 +67,57 @@ def test_purge(self):
...
+        self.api.object_create(self.account, self.container,
+                               obj_name="versioned", data="content1")
+        self.api.object_create(self.account, self.container,
+                               obj_name="versioned", data="content2")
+        oldest_version = check_num_objects_and_get_oldest_version(3)
+
+        self.api.object_create(self.account, self.container,
+                               obj_name="versioned", data="content3")
+        new_oldest_version = check_num_objects_and_get_oldest_version(3)
+        self.assertNotEqual(oldest_version, new_oldest_version)
```
Comment (fvennetier) [inline tests/functional/content/test_content_versioning.py]: I would use `self.assertLess(oldest_version, new_oldest_version)`.

File: meta2v2/meta2_utils.c [hunk 2]
```diff
@@ -940,6 +927,57 @@ _real_delete_and_save_deleted_beans(struct sqlx_sqlite3_s *sq3, GSList *beans,
...
+		header = NULL;
+		beans = NULL;
+
+		err = _alias_fetch_info(sq3, 0, tmp, _search_alias_and_header, &beans);
+		if (!err) {
+			if (cb) {
+				deleted_beans = NULL;
+				err = _real_delete_and_save_deleted_beans(sq3, beans, alias,
+						header, _bean_list_cb, &deleted_beans);
+				cb(u0, deleted_beans);
```
Comment (fvennetier) [inline meta2v2/meta2_utils.c]: We can pass `cb` directly to `_real_delete_and_save_deleted_beans`, don't we?

File: meta2v2/meta2_utils.c [hunk 3]
```diff
@@ -1250,13 +1288,28 @@ static void _extract_chunks_sizes_positions(GSList *beans,
...
+		if ((err = _purge_exceeding_aliases(sq3, max_versions,
+				oio_url_get(url, OIOURL_PATH), _bean_buffer_cb, aliases))) {
+			GRID_WARN("Failed to purge ALIASES: (code=%d) %s",
+					err->code, err->message);
+			_bean_cleanv2(aliases);
+			return err;
+		}
+
+		if (!err)
+			err = _real_delete_aliases(sq3, aliases, cb, u0);
```
Comment (fvennetier) [inline meta2v2/meta2_utils.c]: This can be an `else` statement.

File: conf.json [hunk 4]
```diff
@@ -36,6 +36,11 @@
 			{ "type": "bool", "name": "meta2_flag_precheck_on_generate",
 				"key": "meta2.generate.precheck",
 				"descr": "Should the meta2 check the container state (quota, etc) before generating chunks.",
+				"def": true },
+
+			{ "type": "bool", "name": "meta2_flag_delete_exceeding_versions",
+				"key": "meta2.delete_exceeding_versions",
+				"descr": "When adding alias with versioning, deletes exceeding versions.",
```
Comment (fvennetier) [inline conf.json]: This configuration option is never used.

File: meta2v2/meta2_utils.c (hunk 2 above)
Comment (AymericDu) [reply meta2v2/meta2_utils.c]: `cb` loads a bean list into a list. So `_real_delete_and_save_deleted_beans` creates this bean list.

**Pull request merged**